File: ios/RNMBX/RNMBXMBTilesModule.swift
import Foundation
import MapboxMaps

@objc(RNMBXMBTiles)
class RNMBXMBTilesModule: NSObject {

    private var activeSources: [String: MBTilesSource] = [:]

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    /**
     * Initialize and activate an MBTiles source from a file path
     */
    @objc
    func initMBTilesSource(
        _ filePath: String, sourceId: String, resolver: RCTPromiseResolveBlock,
        rejecter: RCTPromiseRejectBlock
    ) {
        // Handle file URL paths if provided
        var resolvedPath = filePath
        
        // Remove file:// prefix
        if resolvedPath.starts(with: "file://") {
            resolvedPath = resolvedPath.replacingOccurrences(of: "file://", with: "")
        }
        
        // Decode URL encoding (e.g., %20 -> space)
        if let decodedPath = resolvedPath.removingPercentEncoding {
            resolvedPath = decodedPath
        }
        
        RNMBXLogInfo("MBTiles: Attempting to load from path: \(resolvedPath)")

        // Check if file exists
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: resolvedPath) else {
            rejecter("FILE_NOT_FOUND", "MBTiles file not found at path: \(resolvedPath)", nil)
            return
        }

        do {
            // Create and activate the MBTiles source
            let mbSource = try MBTilesSource(
                filePath: resolvedPath, sourceId: sourceId.isEmpty ? nil : sourceId)
            mbSource.activate()
            activeSources[mbSource.id] = mbSource

            // Return source information
            let resultDict: [String: Any] = [
                "id": mbSource.id,
                "url": mbSource.url,
                "isVector": mbSource.isVector,
                "format": mbSource.format,
                "minZoom": mbSource.minZoom as Any,
                "maxZoom": mbSource.maxZoom as Any,
            ]

            resolver(resultDict)
        } catch MBTilesSourceError.couldNotReadFile {
            rejecter("ERROR_READING_FILE", "Could not read the MBTiles file", nil)
        } catch MBTilesSourceError.unsupportedFormat {
            rejecter("UNSUPPORTED_FORMAT", "MBTiles format is not supported", nil)
        } catch {
            rejecter(
                "UNKNOWN_ERROR", "Error initializing MBTiles source: \(error.localizedDescription)",
                nil)
        }
    }

    /**
     * Initialize an MBTiles source from an asset in the app bundle
     */
    @objc
    func initMBTilesSourceFromAsset(
        _ assetName: String, sourceId: String, resolver: RCTPromiseResolveBlock,
        rejecter: RCTPromiseRejectBlock
    ) {
        do {
            // Copy from asset to local file
            let filePath = try MBTilesSource.readAsset(name: assetName)

            // Create and activate the MBTiles source
            let mbSource = try MBTilesSource(
                filePath: filePath, sourceId: sourceId.isEmpty ? nil : sourceId)
            mbSource.activate()
            activeSources[mbSource.id] = mbSource

            // Return source information
            let resultDict: [String: Any] = [
                "id": mbSource.id,
                "url": mbSource.url,
                "isVector": mbSource.isVector,
                "format": mbSource.format,
                "minZoom": mbSource.minZoom as Any,
                "maxZoom": mbSource.maxZoom as Any,
            ]

            resolver(resultDict)
        } catch MBTilesSourceError.couldNotReadFile {
            rejecter("ERROR_READING_FILE", "Could not read the MBTiles asset", nil)
        } catch MBTilesSourceError.unsupportedFormat {
            rejecter("UNSUPPORTED_FORMAT", "MBTiles format is not supported", nil)
        } catch {
            rejecter(
                "UNKNOWN_ERROR",
                "Error initializing MBTiles source from asset: \(error.localizedDescription)", nil)
        }
    }

    /**
     * Initialize an MBTiles source from a remote URL (downloads first)
     */
    @objc
    func initMBTilesSourceFromURL(
        _ urlString: String, sourceId: String, resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard let url = URL(string: urlString) else {
            rejecter("INVALID_URL", "Invalid URL: \(urlString)", nil)
            return
        }

        // Generate a filename from the URL or sourceId
        let fileName = sourceId.isEmpty
            ? url.lastPathComponent
            : "\(sourceId).mbtiles"

        // Get the destination path in the documents directory
        let documentsDirectory = NSSearchPathForDirectoriesInDomains(
            .documentDirectory, .userDomainMask, true)[0]
        let destinationPath = "\(documentsDirectory)/\(fileName)"

        // Download the file
        let task = URLSession.shared.downloadTask(with: url) { [weak self] tempURL, response, error in
            guard let self = self else { return }

            if let error = error {
                rejecter("DOWNLOAD_ERROR", "Failed to download MBTiles file: \(error.localizedDescription)", nil)
                return
            }

            guard let tempURL = tempURL else {
                rejecter("DOWNLOAD_ERROR", "No data received from URL", nil)
                return
            }

            do {
                // Remove existing file if it exists
                let fileManager = FileManager.default
                if fileManager.fileExists(atPath: destinationPath) {
                    try fileManager.removeItem(atPath: destinationPath)
                }

                // Move the downloaded file to the destination
                try fileManager.moveItem(at: tempURL, to: URL(fileURLWithPath: destinationPath))

                // Create and activate the MBTiles source
                let mbSource = try MBTilesSource(
                    filePath: destinationPath, sourceId: sourceId.isEmpty ? nil : sourceId)
                mbSource.activate()
                self.activeSources[mbSource.id] = mbSource

                // Return source information
                let resultDict: [String: Any] = [
                    "id": mbSource.id,
                    "url": mbSource.url,
                    "isVector": mbSource.isVector,
                    "format": mbSource.format,
                    "minZoom": mbSource.minZoom as Any,
                    "maxZoom": mbSource.maxZoom as Any,
                ]

                resolver(resultDict)
            } catch MBTilesSourceError.couldNotReadFile {
                rejecter("ERROR_READING_FILE", "Could not read the downloaded MBTiles file", nil)
            } catch MBTilesSourceError.unsupportedFormat {
                rejecter("UNSUPPORTED_FORMAT", "MBTiles format is not supported", nil)
            } catch {
                rejecter(
                    "UNKNOWN_ERROR",
                    "Error initializing MBTiles source from URL: \(error.localizedDescription)", nil)
            }
        }

        task.resume()
    }

    /**
     * Get the HTTP URL for an active MBTiles source to use in style json
     */
    @objc
    func getMBTilesURL(
        _ sourceId: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock
    ) {
        if let mbSource = activeSources[sourceId] {
            resolver(mbSource.url)
        } else {
            rejecter("SOURCE_NOT_FOUND", "MBTiles source with ID '\(sourceId)' is not active", nil)
        }
    }

    /**
     * Stop and remove an MBTiles source
     */
    @objc
    func removeMBTilesSource(
        _ sourceId: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock
    ) {
        if let mbSource = activeSources[sourceId] {
            mbSource.deactivate()
            activeSources.removeValue(forKey: sourceId)
            resolver(true)
        } else {
            resolver(false)
        }
    }

    /**
     * Check if an MBTiles source is currently active
     */
    @objc
    func isMBTilesSourceActive(
        _ sourceId: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock
    ) {
        resolver(activeSources[sourceId] != nil)
    }

    /**
     * List all active MBTiles sources
     */
    @objc
    func getActiveMBTilesSources(
        _ resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock
    ) {
        let sourceIds = Array(activeSources.keys)
        resolver(sourceIds)
    }

    /**
     * Manually start the MBTiles server
     */
    @objc
    func startServer(_ resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        MBTilesServer.shared.start()
        resolver(MBTilesServer.shared.isRunning)
    }

    /**
     * Manually stop the MBTiles server
     */
    @objc
    func stopServer(_ resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        MBTilesServer.shared.stop()
        resolver(true)
    }

    /**
     * Check if the MBTiles server is running
     */
    @objc
    func isServerRunning(_ resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        resolver(MBTilesServer.shared.isRunning)
    }
}

File: android/src/main/java/com/rnmapbox/rnmbx/modules/RNMBXMBTilesModule.kt
package com.rnmapbox.rnmbx.modules

import android.util.Log
import android.net.Uri
import android.os.Build
import com.facebook.react.bridge.*
import com.rnmapbox.rnmbx.utils.MBTilesServer
import com.rnmapbox.rnmbx.utils.MBTilesSource
import com.rnmapbox.rnmbx.utils.MBTilesSourceException
import java.io.File

class RNMBXMBTilesModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val TAG = "RNMBXMBTilesModule"
    private val activeSources = mutableMapOf<String, MBTilesSource>()

    override fun getName(): String {
        return "RNMBXMBTiles"
    }

    /**
     * Initialize and activate an MBTiles source from a file path
     */
    @ReactMethod
    fun initMBTilesSource(filePath: String, sourceId: String, promise: Promise) {
        try {
            // Note: In v11, we don't need to set Mapbox.isConnected anymore
            // The HTTP requests to localhost will work without this setting

            // Handle Android file paths
            val resolvedPath = if (filePath.startsWith("file://")) {
                Uri.parse(filePath).path ?: filePath.substring(7)
            } else {
                filePath
            }

            // Check if file exists
            val file = File(resolvedPath)
            if (!file.exists()) {
                promise.reject("FILE_NOT_FOUND", "MBTiles file not found at path: $resolvedPath")
                return
            }

            // Create and activate the MBTiles source
            val mbSource = MBTilesSource(resolvedPath, sourceId).apply { activate() }
            activeSources[sourceId] = mbSource

            // Return source information
            val resultMap = Arguments.createMap().apply {
                putString("id", mbSource.id)
                putString("url", mbSource.url)
                putBoolean("isVector", mbSource.isVector)
                putString("format", mbSource.format)
                mbSource.minZoom?.let { putDouble("minZoom", it.toDouble()) }
                mbSource.maxZoom?.let { putDouble("maxZoom", it.toDouble()) }
            }
            promise.resolve(resultMap)

        } catch (e: MBTilesSourceException.CouldNotReadFileException) {
            promise.reject("ERROR_READING_FILE", "Could not read the MBTiles file")
        } catch (e: MBTilesSourceException.UnsupportedFormatException) {
            promise.reject("UNSUPPORTED_FORMAT", "MBTiles format is not supported")
        } catch (e: Exception) {
            promise.reject("UNKNOWN_ERROR", "Error initializing MBTiles source: ${e.localizedMessage}")
        }
    }

    /**
     * Initialize an MBTiles source from an asset in the app bundle
     */
    @ReactMethod
    fun initMBTilesSourceFromAsset(assetName: String, sourceId: String, promise: Promise) {
        try {
            // Note: In v11, we don't need to set Mapbox.isConnected anymore
            // The HTTP requests to localhost will work without this setting

            // Copy from asset to local file
            val filePath = MBTilesSource.readAsset(reactContext, assetName)

            // Create and activate the MBTiles source
            val mbSource = MBTilesSource(filePath, sourceId).apply { activate() }
            activeSources[sourceId] = mbSource

            // Return source information
            val resultMap = Arguments.createMap().apply {
                putString("id", mbSource.id)
                putString("url", mbSource.url)
                putBoolean("isVector", mbSource.isVector)
                putString("format", mbSource.format)
                mbSource.minZoom?.let { putDouble("minZoom", it.toDouble()) }
                mbSource.maxZoom?.let { putDouble("maxZoom", it.toDouble()) }
            }
            promise.resolve(resultMap)

        } catch (e: MBTilesSourceException.CouldNotReadFileException) {
            promise.reject("ERROR_READING_FILE", "Could not read the MBTiles asset")
        } catch (e: MBTilesSourceException.UnsupportedFormatException) {
            promise.reject("UNSUPPORTED_FORMAT", "MBTiles format is not supported")
        } catch (e: Exception) {
            promise.reject("UNKNOWN_ERROR", "Error initializing MBTiles source from asset: ${e.localizedMessage}")
        }
    }

    /**
     * Initialize an MBTiles source from a remote URL (downloads first)
     */
    @ReactMethod
    fun initMBTilesSourceFromURL(urlString: String, sourceId: String, promise: Promise) {
        Thread {
            try {
                val url = java.net.URL(urlString)

                // Generate a filename from the URL or sourceId
                val fileName = if (sourceId.isEmpty()) {
                    url.path.substringAfterLast("/")
                } else {
                    "$sourceId.mbtiles"
                }

                // Get the destination path
                val destinationFile = File(reactContext.filesDir, fileName)

                // Download the file
                url.openStream().use { input ->
                    java.io.FileOutputStream(destinationFile).use { output ->
                        input.copyTo(output)
                    }
                }

                // Create and activate the MBTiles source
                val effectiveSourceId = if (sourceId.isEmpty()) {
                    fileName.substringBefore(".")
                } else {
                    sourceId
                }
                val mbSource = MBTilesSource(destinationFile.absolutePath, effectiveSourceId).apply { activate() }
                activeSources[effectiveSourceId] = mbSource

                // Return source information
                val resultMap = Arguments.createMap().apply {
                    putString("id", mbSource.id)
                    putString("url", mbSource.url)
                    putBoolean("isVector", mbSource.isVector)
                    putString("format", mbSource.format)
                    mbSource.minZoom?.let { putDouble("minZoom", it.toDouble()) }
                    mbSource.maxZoom?.let { putDouble("maxZoom", it.toDouble()) }
                }
                promise.resolve(resultMap)

            } catch (e: MBTilesSourceException.CouldNotReadFileException) {
                promise.reject("ERROR_READING_FILE", "Could not read the downloaded MBTiles file")
            } catch (e: MBTilesSourceException.UnsupportedFormatException) {
                promise.reject("UNSUPPORTED_FORMAT", "MBTiles format is not supported")
            } catch (e: java.net.MalformedURLException) {
                promise.reject("INVALID_URL", "Invalid URL: $urlString")
            } catch (e: java.io.IOException) {
                promise.reject("DOWNLOAD_ERROR", "Failed to download MBTiles file: ${e.localizedMessage}")
            } catch (e: Exception) {
                promise.reject("UNKNOWN_ERROR", "Error initializing MBTiles source from URL: ${e.localizedMessage}")
            }
        }.start()
    }

    /**
     * Get the HTTP URL for an active MBTiles source to use in style json
     */
    @ReactMethod
    fun getMBTilesURL(sourceId: String, promise: Promise) {
        val mbSource = activeSources[sourceId]
        if (mbSource != null) {
            promise.resolve(mbSource.url)
        } else {
            promise.reject("SOURCE_NOT_FOUND", "MBTiles source with ID '$sourceId' is not active")
        }
    }

    /**
     * Stop and remove an MBTiles source
     */
    @ReactMethod
    fun removeMBTilesSource(sourceId: String, promise: Promise) {
        val mbSource = activeSources[sourceId]
        if (mbSource != null) {
            mbSource.deactivate()
            activeSources.remove(sourceId)
            promise.resolve(true)
        } else {
            promise.resolve(false)
        }
    }

    /**
     * Check if an MBTiles source is currently active
     */
    @ReactMethod
    fun isMBTilesSourceActive(sourceId: String, promise: Promise) {
        promise.resolve(activeSources.containsKey(sourceId))
    }

    /**
     * List all active MBTiles sources
     */
    @ReactMethod
    fun getActiveMBTilesSources(promise: Promise) {
        val sources = Arguments.createArray()
        activeSources.forEach { (id, _) ->
            sources.pushString(id)
        }
        promise.resolve(sources)
    }

    /**
     * Manually start the MBTiles server
     */
    @ReactMethod
    fun startServer(promise: Promise) {
        try {
            MBTilesServer.start()
            promise.resolve(MBTilesServer.isRunning)
        } catch (e: Exception) {
            promise.reject("SERVER_ERROR", "Error starting MBTiles server: ${e.localizedMessage}")
        }
    }

    /**
     * Manually stop the MBTiles server
     */
    @ReactMethod
    fun stopServer(promise: Promise) {
        try {
            MBTilesServer.stop()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("SERVER_ERROR", "Error stopping MBTiles server: ${e.localizedMessage}")
        }
    }

    /**
     * Check if the MBTiles server is running
     */
    @ReactMethod
    fun isServerRunning(promise: Promise) {
        promise.resolve(MBTilesServer.isRunning)
    }

    @ReactMethod
    fun addListener(eventName: String) {
        // Required for RN built-in Event Emitter Calls
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // Required for RN built-in Event Emitter Calls
    }

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        // Cleanup when React context is destroyed
        activeSources.forEach { (_, source) ->
            source.deactivate()
        }
        activeSources.clear()
    }
}
